scripts/3.0-migration.sh

```diff
@@ -0,0 +1,252 @@
+done
+for ((i = 0; i < ${#user_account_resource_names[@]}; i++));do
+    terraform state show -no-color ${user_account_resource_names[$i]} | grep -v "   user_account_id" | grep -v "   id " >> cyral_migration_repository_access_rules_and_user_accounts.txt
+done
+
+mv cyral_migration_repository_access_rules_and_user_accounts.txt cyral_migration_repository_access_rules_and_user_accounts.tf
+terraform fmt
+
+echo; echo; echo;
+echo "Now that the Terraform state is up-to-date, let's clean up your .tf file."
+echo "This script created a new .tf file containing the resources definitions"
+echo "for the cyral_repository_access_rules and cyral_repository_user_accounts"
+echo "that were migrated into your Terraform state. The new .tf file is called:"
+echo
+echo "cyral_migration_repository_access_rules_and_user_accounts.tf"
+echo
+echo
+echo "It is finally time to remove the old resources from your .tf files."
+echo "Please perform the following actions: "
+echo
+echo "  1.  Find all references to cyral_repository_identity_maps and"
+echo "      cyral_repository_local_accounts in your .tf file and remove the"
+echo "      entire resource defintion for each one."
+echo "  2.  Remove the empty resource definitions for the "
+echo "      cyral_repository_access_rules and cyral_repository_user_account"
+echo "      that were added to the end of your .tf file, which is named:"
+echo "      ${CYRAL_TF_FILE_PATH}"
+echo
+echo
+echo "When you are done, run the following command:"
+echo "terraform plan"
+echo
+echo
+echo "If migration was successful, you should see the following message:"
+echo "No changes. Your infrastructure matches the configuration."
+echo
+echo "If migration was not successful, you have to option to revert to your"
+echo "previous Terraform state and try again."
+echo
+read -p "Would you revert to your previous state and try the migration again? [N/y] " -n 1 -r
+echo    # move to a new line
+if [[ ! $REPLY =~ ^[Yy]$ ]]
+then
+    echo "Exiting..."
+    [[ "$0" = "$BASH_SOURCE" ]] && exit 1 || return 1
+fi
+
+echo "Your previous .tf file was copied before the migration was ran. It is called "
+echo "cyral_terraform_migration_backup_configuration.txt"
+echo
+echo "Please perform the following action before proceding:"
+echo "  1.  Replace the contents of your .tf file ${CYRAL_TF_FILE_PATH} "
+echo "      with the contents of cyral_terraform_migration_backup_configuration.txt. "
+echo "  2.  Delete the following files that were created by the script: "
+echo "      - cyral_terraform_migration_backup_configuration.txt"
+echo "      - cyral_migration_repository_access_rules_and_user_accounts.tf"
+echo
+read -p "Are you ready to revert to your pre-migration Terraform state? [N/y] " -n 1 -r
+echo    # move to a new line
+if [[ ! $REPLY =~ ^[Yy]$ ]]
+then
+    echo "Exiting..."
+    [[ "$0" = "$BASH_SOURCE" ]] && exit 1 || return 1
+fi
+
+# Downgrade Terraform
+terraform init -upgrade
```

Comment:
This would only downgrade the terraform provider version if users change back the version constraint on the .tf file, so we might want to add that instruction before running this command.

Reply:
Oh, I just noticed that the backup file for the .tf config will have the old version and when users replace the .tf with the backup content the downgrade will always work, please ignore the comment above :)